pkg/controller/nodeipam/ipam/multicidrset/multi_cidr_set.go

```diff
@@ -0,0 +1,365 @@
+	return fmt.Sprintf("Creation of New CIDR Set failed for %s. "+
+		"PerNodeMaskSize %d is too big for CIDR Mask %d, Maximum difference allowed "+
+		"is %d", err.cidr, err.subnetMaskSize, err.clusterMaskSize, clusterSubnetMaxDiff)
+}
+
+// NewMultiCIDRSet creates a new MultiCIDRSet.
+func NewMultiCIDRSet(cidrConfig *net.IPNet, perNodeHostBits int) (*MultiCIDRSet, error) {
+	clusterMask := cidrConfig.Mask
+	clusterMaskSize, bits := clusterMask.Size()
+
+	var subNetMaskSize int
+	switch /*v4 or v6*/ {
+	case cidrConfig.IP.To4() != nil:
+		{
+			subNetMaskSize = 32 - perNodeHostBits
+		}
+	case cidrConfig.IP.To16() != nil:
+		{
+			subNetMaskSize = 128 - perNodeHostBits
+		}
+	}
+
+	if (cidrConfig.IP.To4() == nil) && (subNetMaskSize-clusterMaskSize > clusterSubnetMaxDiff) {
+		return nil, &CIDRSetSubNetTooBigErr{
+			cidr:            cidrConfig.String(),
+			subnetMaskSize:  subNetMaskSize,
+			clusterMaskSize: clusterMaskSize,
+		}
+	}
+
+	// Register MultiCIDRSet metrics.
+	registerCidrsetMetrics()
+
+	return &MultiCIDRSet{
+		ClusterCIDR:      cidrConfig,
+		nodeMask:         net.CIDRMask(subNetMaskSize, bits),
+		clusterMaskSize:  clusterMaskSize,
+		MaxCIDRs:         1 << uint32(subNetMaskSize-clusterMaskSize),
+		NodeMaskSize:     subNetMaskSize,
+		Label:            cidrConfig.String(),
+		AllocatedCIDRMap: make(map[string]bool, 0),
+	}, nil
+}
+
+func (s *MultiCIDRSet) indexToCIDRBlock(index int) (*net.IPNet, error) {
+	var ip []byte
+	switch /*v4 or v6*/ {
+	case s.ClusterCIDR.IP.To4() != nil:
+		{
+			j := uint32(index) << uint32(32-s.NodeMaskSize)
+			ipInt := (binary.BigEndian.Uint32(s.ClusterCIDR.IP)) | j
+			ip = make([]byte, net.IPv4len)
+			binary.BigEndian.PutUint32(ip, ipInt)
+		}
+	case s.ClusterCIDR.IP.To16() != nil:
+		{
+			// leftClusterIP      |     rightClusterIP
+			// 2001:0DB8:1234:0000:0000:0000:0000:0000
+			const v6NBits = 128
+			const halfV6NBits = v6NBits / 2
+			leftClusterIP := binary.BigEndian.Uint64(s.ClusterCIDR.IP[:halfIPv6Len])
+			rightClusterIP := binary.BigEndian.Uint64(s.ClusterCIDR.IP[halfIPv6Len:])
+
+			ip = make([]byte, net.IPv6len)
+
+			if s.NodeMaskSize <= halfV6NBits {
+				// We only care about left side IP.
+				leftClusterIP |= uint64(index) << uint(halfV6NBits-s.NodeMaskSize)
+			} else {
+				if s.clusterMaskSize < halfV6NBits {
+					// see how many bits are needed to reach the left side.
+					btl := uint(s.NodeMaskSize - halfV6NBits)
+					indexMaxBit := uint(64 - bits.LeadingZeros64(uint64(index)))
+					if indexMaxBit > btl {
+						leftClusterIP |= uint64(index) >> btl
+					}
+				}
+				// the right side will be calculated the same way either the
+				// subNetMaskSize affects both left and right sides.
+				rightClusterIP |= uint64(index) << uint(v6NBits-s.NodeMaskSize)
+			}
+			binary.BigEndian.PutUint64(ip[:halfIPv6Len], leftClusterIP)
+			binary.BigEndian.PutUint64(ip[halfIPv6Len:], rightClusterIP)
+		}
+	default:
+		return nil, fmt.Errorf("invalid IP: %s", s.ClusterCIDR.IP)
```

Comment:
can this happen?
I think that this controller always assume CIDRs were validated

Reply:
It doesn't hurt to handle the error rather than fall off the end with uninitialized values.